### PR TITLE
Allow request whitelisting in CsrfProtectionMiddleware.

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

**Comment by dereuromark** [comment]
Your descriptions get more verbose every time :P 

Jokes aside: There seem to be test fails due to some invalid tokens (PHP7 vs PHP5?)

File: tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php [hunk 1]
```diff
@@ -309,4 +310,24 @@ public function testConfigurationValidate()
...
+        $request = new ServerRequest([
+            'environment' => [
+                'REQUEST_METHOD' => 'POST',
+                'ALLOW_ME' => 'let-me-pass'
+            ],
+        ]);
+        $response = new Response();
+
+        $middleware = new CsrfProtectionMiddleware();
+        $middleware->whitelistCallback(function (ServerRequestInterface $request): ?bool {
```
Comment (dereuromark) [inline tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php]: No return types for PHP5 :)

Comment (ADmad) [reply tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php]: Doh, I should have made PR for 4.x and let someone else backport it :P

**Comment by ADmad** [comment]
> Your descriptions get more verbose every time :P

The code speaks for itself :loudspeaker: :slightly_smiling_face:.

**Comment by ADmad** [comment]
I'll make a PR to the docs once this is approved.

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 2]
```diff
@@ -66,6 +66,13 @@ class CsrfProtectionMiddleware
      */
     protected $_config = [];
 
+    /**
+     * Callback for allowing to skip token check for particular request.
```
Comment (markstory) [inline src/Http/Middleware/CsrfProtectionMiddleware.php]:
```suggestion
     * Callback for deciding whether or not to skip the token check for particular request.
```

Comment (markstory) [reply src/Http/Middleware/CsrfProtectionMiddleware.php]: Might also want to mention that returning true from this callback will skip the csrf check.

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 3]
```diff
@@ -108,6 +121,22 @@ public function __invoke(ServerRequest $request, Response $response, $next)
         return $next($request, $response);
     }
 
+    /**
+     * Set callback for allowing to skip token check for particular request.
+     *
+     * The callback will receive request instance as argument and must return
+     * `true` if you want to skip token check for the particular request.
```
Comment (markstory) [inline src/Http/Middleware/CsrfProtectionMiddleware.php]:
```suggestion
     * `true` if you want to skip token check for the current request.
```

File: tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php [hunk 4]
```diff
@@ -309,4 +310,24 @@ public function testConfigurationValidate()
...
+                'REQUEST_METHOD' => 'POST',
+                'ALLOW_ME' => 'let-me-pass'
+            ],
+        ]);
+        $response = new Response();
+
+        $middleware = new CsrfProtectionMiddleware();
+        $middleware->whitelistCallback(function (ServerRequestInterface $request) {
+            return !empty($request->getServerParams()['ALLOW_ME']);
+        });
```
Comment (markstory) [inline tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php]: Couldn't you assert that the next middleware layer is called?

**Comment by ADmad** [comment]
Updates done as per feedback.

**Comment by markstory** [comment]
Nice work @ADmad

**Pull request merged**